From: Sebastian Wick <sebastian.wick@redhat.com>
Date: Mon, 12 Jan 2026 17:38:02 +0100
Subject: utils: Do not follow symlinks in flatpak_open_file_at

We use flatpak_open_file_at in the context of the system helper to open
files written by a user. This means that we want to prevent DOS and
exposing files which only the system helper has access to.

To prevent DOS and avoid side-effects, the file is opened with
O_NONBLOCK and O_NOCTTY.

To prevent leaking files, the file is supposed to not open symlinks.
This part, we failed at. We check if the opened file is a regular file,
but what we actually checked is, if the file a symlink might point at is
a regular file.

Fix this by also specifying O_NOFOLLOW in openat.

Bug: https://github.com/flatpak/flatpak/security/advisories/GHSA-2fxp-43j9-pwvc
---
 common/flatpak-utils.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/common/flatpak-utils.c b/common/flatpak-utils.c
index e0f50bb..30c7b00 100644
--- a/common/flatpak-utils.c
+++ b/common/flatpak-utils.c
@@ -1912,7 +1912,7 @@ flatpak_allocate_tmpdir (int           tmpdir_dfd,
 }
 
 /* Carefully opens a file from a base directory and subpath,
- * making sure that its not a symlink.
+ * making sure that its not a symlink, pipe, etc.
  */
 int
 flatpak_open_file_at (int           dfd,
@@ -1925,7 +1925,7 @@ flatpak_open_file_at (int           dfd,
   struct stat tmp_st_buf;
 
   do
-    fd = openat (dfd, subpath, O_RDONLY | O_NONBLOCK | O_CLOEXEC | O_NOCTTY);
+    fd = openat (dfd, subpath, O_NOFOLLOW | O_RDONLY | O_NONBLOCK | O_CLOEXEC | O_NOCTTY);
   while (G_UNLIKELY (fd == -1 && errno == EINTR));
   if (fd == -1)
     {
